#http://retractiondatabase.org/RetractionSearch.aspx?

Anlyse infomrmation about papers and retractions directly taken from this webpage. 
Type: article or letter.
Reasons: Error in data, error in results and/or conclusions


In [1]:
import os
os.getcwd()
path = "/project/eLife_retractions"
os.chdir(path)

In [2]:
import pandas as pd

df =pd.read_csv('retractions600.txt', sep = '\t')

df = df.reset_index()
df

level_0  \
0          600 Items Displayed Out of 791 Item(s) Found   
1     Structure revision of aspergicin by the crysta...   
2     (PHY) Chemistry; (PHY) Crystallography/Spectro...   
3        Natural Product Research ---Taylor and Francis   
4     Department of Chemical Engineering, Foshan Uni...   
5     http://retractionwatch.com/2018/01/16/caught-n...   
6                                                   NaN   
7                                                   NaN   
8     Solar photothermochemical alkane reverse combu...   
9                        (PHY) Chemistry; (PHY) Energy;   
10    Proceedings of the National Academy of Science...   
11    Department of Chemistry and Biochemistry, Univ...   
12    Car safety seats for children: rear facing for...   
13    (B/T) Transportation; (HSC) Medicine - Pediatr...   
14                  Injury Prevention ---BMJ Publishing   
15    Center for Applied Biomechanics, University of...   
16    http://retractionwatch.com/2017/12/15/saving-k...   
17                                                  NaN   
18                                                  NaN   
19    Car safety seats for children: rear facing for...   
20    (B/T) Transportation; (HSC) Medicine - Pediatr...   
21                  Injury Prevention ---BMJ Publishing   
22    Center for Applied Biomechanics, University of...   
23    http://retractionwatch.com/2017/12/15/saving-k...   
24                                                  NaN   
25                                                  NaN   
26    Transgenic mice overexpressing the ALS-linked ...   
27    (BLS) Biology - Cellular; (BLS) Biology - Mole...   
28    Acta Neuropathologica Communications ---Spring...   
29    Center for Translational Research in Neurodege...   
...                                                 ...   
4032                                                NaN   
4033                                                NaN   
4034                                                NaN   
4035                                                NaN   
4036                                                NaN   
4037  Expression of matrix metalloproteinase-26 (MMP...   
4038  (BLS) Anatomy/Physiology; (BLS) Biology - Cell...   
4039                          Life Sciences ---Elsevier   
4040  State Key Laboratory of Reproductive Biology, ...   
4041                                                NaN   
4042                                                NaN   
4043                                                NaN   
4044                                                NaN   
4045  [NEt4][AuCl2][AuCl3]: Solid-State Evidence of ...   
4046  (PHY) Chemistry; (PHY) Crystallography/Spectro...   
4047                         Angewandte Chemie ---Wiley   
4048  School of Chemistry, University of Manchester,...   
4049  Origin of the Obliquities of the Giant Planets...   
4050  (PHY) Space - Astronomy; (PHY) Space - Astroph...   
4051                  Nature ---Nature Publishing Group   
4052  Facultad de Ciencias Astron�micas y Geof�sicas...   
4053  Diffractive electron imaging of nanoparticles ...   
4054          (PHY) Chemistry; (PHY) Materials Science;   
4055  Nature Materials ---Springer - Nature Publishi...   
4056  Department of Physics and Astronomy, Arizona S...   
4057  YRKL sequence of influenza virus M1 functions ...   
4058  (BLS) Biology - Cellular; (BLS) Genetics; (HSC...   
4059  Journal of Virology ---American Society for Mi...   
4060  Department of Microbiology, Immunology and Mol...   
4061                                                NaN   

                                    level_1                 level_2  \
0                                       NaN                     NaN   
1      +Error in Results and/or Conclusions                Feng Zhu   
2                                       NaN              Jie-sen Li   
3                                       NaN             Wu-chen Xie   
4                                 

In [3]:
df2 = df[['level_0','Date/PubMedID/DOI', 'Date/PubMedID/DOI.1']]
df2 = df2.dropna(subset = ['Date/PubMedID/DOI'])
df2 = df2.dropna().reset_index(drop = 'True')
df2

level_0  \
0     Structure revision of aspergicin by the crysta...   
1     (PHY) Chemistry; (PHY) Crystallography/Spectro...   
2        Natural Product Research ---Taylor and Francis   
3     Solar photothermochemical alkane reverse combu...   
4                        (PHY) Chemistry; (PHY) Energy;   
5     Proceedings of the National Academy of Science...   
6     Car safety seats for children: rear facing for...   
7     (B/T) Transportation; (HSC) Medicine - Pediatr...   
8                   Injury Prevention ---BMJ Publishing   
9     Car safety seats for children: rear facing for...   
10    (B/T) Transportation; (HSC) Medicine - Pediatr...   
11                  Injury Prevention ---BMJ Publishing   
12    Transgenic mice overexpressing the ALS-linked ...   
13    (BLS) Biology - Cellular; (BLS) Biology - Mole...   
14    Acta Neuropathologica Communications ---Spring...   
15    Schisandrin B suppresses NLRP3 inflammasome ac...   
16    (BLS) Biochemistry; (BLS) Biology - Cellular; ...   
17    Biochemical and Biophysical Research Communica...   
18    Transition-Metal-Free C-CN/C-H Cross-Coupling:...   
19                                     (PHY) Chemistry;   
20    Organic Letters ---American Chemical Society (...   
21    NAD(P)H:quinone oxidoreductase 1 protects lung...   
22    (BLS) Biochemistry; (BLS) Biology - Cellular; ...   
23          Free Radical Biology & Medicine ---Elsevier   
24    Effects of synthetic magnetic flux in a parity...   
25                                       (PHY) Physics;   
26       Physical Review A ---American Physical Society   
27    Whole-Genome De Novo Sequencing of the Lignin-...   
28    (BLS) Genetics; (BLS) Microbiology; (BLS) Zool...   
29    Genome Announcements ---American Society for M...   
...                                                 ...   
1608  Structure of MsbA from E. coli: a homolog of t...   
1609  (BLS) Biochemistry; (BLS) Biology - Molecular;...   
1610  Science ---American Association for the Advanc...   
1611  Structure of the ABC transporter MsbA in compl...   
1612  (BLS) Biochemistry; (BLS) Biology - Molecular;...   
1613  Science ---American Association for the Advanc...   
1614  X-ray structure of the EmrE multidrug transpor...   
1615  (BLS) Biology - Molecular; (BLS) Microbiology;...   
1616  Science ---American Association for the Advanc...   
1617  Dual signaling regulated by calcyon, a D1 dopa...   
1618  (BLS) Biochemistry; (BLS) Biology - Molecular;...   
1619  Science ---American Association for the Advanc...   
1620  Silencing of B cell receptor signals in human ...   
1621                       (HSC) Medicine - Immunology;   
1622  The Journal of Experimental Medicine ---Rockef...   
1623  Expression of matrix metalloproteinase-26 (MMP...   
1624  (BLS) Anatomy/Physiology; (BLS) Biology - Cell...   
1625                          Life Sciences ---Elsevier   
1626  [NEt4][AuCl2][AuCl3]: Solid-State Evidence of ...   
1627  (PHY) Chemistry; (PHY) Crystallography/Spectro...   
1628                         Angewandte Chemie ---Wiley   
1629  Origin of the Obliquities of the Giant Planets...   
1630  (PHY) Space - Astronomy; (PHY) Space - Astroph...   
1631                  Nature ---Nature Publishing Group   
1632  Diffractive electron imaging of nanoparticles ...   
1633          (PHY) Chemistry; (PHY) Materials Science;   
1634  Nature Materials ---Springer - Nature Publishi...   
1635  YRKL sequence of influenza virus M1 functions ...   
1636  (BLS) Biology - Cellular; (BLS) Genetics; (HSC...   
1637  Journal of Virology ---American Society for Mi...   

                        Date/PubMedID/DOI                  Date/PubMedID/DOI.1  
0                               3/13/2017                             1/8/2018  
1                                28287288                             29310452  
2           10.1080/14786419.2017.1300800        10.1080/14786419.2017.1422331  
3                                3/8/2016                        

In [4]:
df2.shape[0] % 3

0

In [5]:
from datetime import datetime
from dateutil.parser import parse
import pandas as pd

DOI_article = []
PubMed_ID_article = []
DOI_retraction = []
PubMed_ID_retraction = []
#Journal = []
date_article = []
date_retraction = []

for index, row in df2.iterrows():
    if len(row['Date/PubMedID/DOI']) > 10: 
        try:
            date_article.append(datetime.strptime(df2.iloc[index -2, 1], '%m/%d/%Y'))
            date_retraction.append(datetime.strptime(df2.iloc[index -2, 2], '%m/%d/%Y'))

            DOI_article.append(row['Date/PubMedID/DOI'])
            DOI_retraction.append(row['Date/PubMedID/DOI.1'])
        
            PubMed_ID_article.append(df2.iloc[index -1, 1])
            PubMed_ID_retraction.append(df2.iloc[index -1, 2])
            
        except ValueError:
            pass
                      
        
        #print(row['Date/PubMedID/DOI'],row['Date/PubMedID/DOI.1'])

# df2 = pd.DataFrame({'Journal': Journal,'DOI':DOI_a
df_retracted_retraction = pd.DataFrame({'doi':DOI_article, 
                    'pmid_article':PubMed_ID_article, 
                    'doi_retraction':DOI_retraction, 
                    'pmid_retraction':PubMed_ID_retraction,
                    'date_article': date_article,
                    'date_retraction' : date_retraction}).drop_duplicates()

#mask = df_retracted_retraction.date_retraction != 'unavailable'
#mask4 = df_retracted_retraction.date_retraction != "0"
#df_retracted_retraction = df_retracted_retraction[mask3 & mask4]
df_retracted_retraction



date_article date_retraction                                  doi  \
0     2017-03-13      2018-01-08        10.1080/14786419.2017.1300800   
1     2016-03-08      2018-01-03              10.1073/pnas.1516945113   
2     2007-12-01      2017-12-20               10.1136/ip.2006.015115   
3     2007-12-01      2017-06-30               10.1136/ip.2006.015115   
4     2016-11-18      2017-12-13            10.1186/s40478-016-0393-5   
5     2017-08-31      2017-12-09           10.1016/j.bbrc.2017.08.079   
6     2017-09-11      2017-11-30          10.1021/acs.orglett.7b02473   
7     2012-02-01      2017-11-20  10.1016/j.freeradbiomed.2011.11.027   
8     2017-10-10      2017-11-16           10.1103/PhysRevA.96.043821   
9     2017-08-10      2017-11-16             10.1128/genomeA.00731-17   
10    2017-05-03      2017-11-13                   10.1039/C7TA02103H   
11    2014-12-19      2017-10-26        10.1126/science.346.6216.1486   
12    2017-10-01      2017-10-25            10.21873/anticanres.11969   
13    2012-10-01      2017-10-20      10.1001/archpediatrics.2012.999   
14    2012-10-01      2017-09-21      10.1001/archpediatrics.2012.999   
15    2008-01-01      2017-10-14                   10.1093/ndt/gfm690   
16    2016-09-20      2017-10-03              10.1021/acs.joc.6b02086   
17    2016-05-11      2017-10-02        10.1158/1535-7163.MCT-15-0375   
18    2016-09-09      2017-10-01                          unavailable   
19    2017-07-25      2017-09-28                10.3892/mmr.2017.7084   
20    2014-12-02      2017-09-26                 10.3892/or.2014.3644   
21    2014-12-26      2017-09-26         10.3109/19401736.2014.987245   
22    2017-04-18      2017-09-19               10.1001/jama.2017.1469   
23    2017-05-11      2017-09-15            10.1186/s13033-017-0143-5   
24    2017-02-23      2017-09-14         10.1021/acs.jmedchem.6b01417   
25    2017-01-25      2017-09-08                    10.7150/jca.15753   
26    2017-07-03      2017-09-06                   10.1039/c7ew00132k   
27    2017-01-31      2017-09-05               10.1038/sdata.2016.120   
28    2017-07-05      2017-09-05                10.1139/cjp-2017-0290   
29    2017-07-08      2017-09-05           10.1016/j.gene.2017.07.018   
..           ...             ...                                  ...   
421   2005-10-03      2008-04-17            10.1016/j.jep.2005.08.060   
422   2005-04-08      2008-04-17            10.1016/j.jep.2005.01.027   
423   2004-12-17      2008-04-11              10.1126/science.1103776   
424   2001-11-08      2008-03-06                     10.1038/35102506   
425   2004-07-06      2008-02-20              10.1073/pnas.0403155101   
426   2006-02-08      2008-02-14            10.1016/j.jmb.2006.01.087   
427   2006-07-06      2008-01-31                  10.1038/nature04836   
428   2007-12-11      2008-01-05              10.2310/6650.2007.00027   
429   2001-07-01      2007-12-15              10.1182/blood.V98.1.125   
430   2007-06-01      2007-12-01           10.1152/ajpendo.00702.2006   
431   2004-08-05      2007-10-31             10.1038/sj.emboj.7600343   
432   1998-02-06      2007-10-11                    10.1021/bi972113+   
433   2004-02-17      2007-10-09            10.1016/j.cub.2004.02.003   
434   2007-08-02      2007-10-01           10.1016/j.epsl.2007.07.046   
435   2002-06-21      2007-08-10              10.1126/science.1071776   
436   2007-03-16      2007-07-20         10.1371/journal.pcbi.0030051   
437   2006-07-01      2007-07-01           10.1016/j.aaen.2006.05.005   
438   2006-08-01      2007-06-13                    10.1021/bi060547b   
439   2005-06-10      2007-06-13                    10.1021/bi050786j   
440   2004-06-01      2007-02-01                10.1104/pp.103.037770   
441   2001-09-07      2006-12-22        10.1126/science.293.5536.1793   
442   2005-05-13      2006-12-22              10.1126/science.1107733   
443   2005-12-23      2006-12-22              10.1126/science.

# Let's use citation_context to get all the manuscripts citing these dois from PMC OAS

In [6]:
df_retracted_retraction.doi[445:]

445               10.1084/jem.20020881
446          10.1016/j.lfs.2005.05.045
447             10.1002/anie.200602092
448                10.1038/nature04577
449                   10.1038/nmat1531
450    10.1128/JVI.80.5.2291-2308.2006
Name: doi, dtype: object

In [7]:
# using get_papers notebook and module/bs_preprocess.py from citation_context

DOI_list = df_retracted_retraction.doi[400:]
ds_name = 'df_retracted_retraction_400_450'


%store DOI_list
%store ds_name

%run get_papers.ipynb

Stored 'DOI_list' (Series)
Stored 'ds_name' (str)
10.1128/IAI.69.8.4858-4869.2001
data/FTP_PUBMED_papers/comm_use.A-B.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.C-H.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.I-N.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.O-Z.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.A-B.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.C-H.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.I-N.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.O-Z.xml.tar.gz
No articles found citing  10.1128/IAI.69.8.4858-4869.2001
Data saved to data/analysis/df_retracted_retraction_400_450.tsv
10.1104/pp.011019
data/FTP_PUBMED_papers/comm_use.A-B.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.C-H.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.I-N.xml.tar.gz
data/FTP_PUBMED_papers/comm_use.O-Z.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.A-B.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.C-H.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.I-N.xml.tar.gz
data/FTP_PUBMED_papers/non_comm_use.O-Z.xml.t

KeyboardInterrupt: 

#### Let's read the files with the xml text.

In [40]:
import os
import pandas as pd
import glob 



data_path = 'data'
analysis_path = os.path.join(data_path, 'analysis')

df_articles_citing_retracted_paper = pd.DataFrame()


for filename in sorted(glob.glob(analysis_path + '/df_*0.tsv')):
    print(filename)

    df_in_file = pd.read_csv(filename, sep='\t', encoding='utf-8')
    print(df_in_file.index[-1])
    
    
    df_articles_citing_retracted_paper = df_articles_citing_retracted_paper.append(df_in_file)
    
df_articles_citing_retracted_paper.reset_index(drop = True)


data/analysis/df_retracted_retraction_0_100.tsv
68
data/analysis/df_retracted_retraction_100_200.tsv
93
data/analysis/df_retracted_retraction_200_300.tsv
84
data/analysis/df_retracted_retraction_300_400.tsv
127
data/analysis/df_retracted_retraction_400_450.tsv
3


DOI_cited  \
0                 10.1136/ip.2006.015115   
1                 10.1136/ip.2006.015115   
2              10.1186/s40478-016-0393-5   
3    10.1016/j.freeradbiomed.2011.11.027   
4               10.1128/genomeA.00731-17   
5          10.1126/science.346.6216.1486   
6                  10.3892/mmr.2017.7084   
7              10.1186/s13033-017-0143-5   
8              10.1186/s13033-017-0143-5   
9                      10.7150/jca.15753   
10                10.1038/sdata.2016.120   
11                    10.1038/ncomms6446   
12                    10.1038/ncomms6446   
13                    10.1093/nar/gks812   
14                 10.3892/mmr.2016.5294   
15                    10.1002/cbin.10615   
16                    10.1002/cbin.10291   
17                    10.1002/cbin.10291   
18          10.1371/journal.pone.0164378   
19                  10.1021/jacs.6b01652   
20          10.1371/journal.pone.0164571   
21          10.1371/journal.pone.0164571   
22               10.1186/1472-6882-14-82   
23               10.1186/1472-6882-14-82   
24                  10.1155/2017/6258146   
25                 10.1104/pp.110.157180   
26                   10.1289/ehp.1408282   
27                   10.1289/ehp.1408282   
28             10.1186/s12889-017-4082-5   
29          10.1371/journal.pone.0121549   
..                                   ...   
350             10.1200/JCO.2007.11.0593   
351             10.1200/JCO.2007.11.0593   
352             10.1200/JCO.2007.11.0593   
353             10.1200/JCO.2007.11.0593   
354              10.1073/pnas.0809516106   
355              10.1073/pnas.0809516106   
356        10.1158/0008-5472.CAN-07-5937   
357        10.1158/0008-5472.CAN-07-5937   
358        10.1158/0008-5472.CAN-03-3025   
359                  10.1038/nature02356   
360                  10.1038/nature02356   
361              10.1126/science.1124755   
362              10.1126/science.1124755   
363              10.1126/science.1124755   
364              10.1126/science.1124755   
365              10.1126/science.1124755   
366           10.1016/j.cell.2009.01.054   
367        10.1158/0008-5472.CAN-04-4194   
368        10.1158/0008-5472.CAN-04-4194   
369        10.1158/0008-5472.CAN-04-4194   
370        10.1158/0008-5472.CAN-04-4194   
371              10.1073/pnas.0503877102   
372                10.1083/jcb.200705182   
373                10.1083/jcb.200705182   
374                10.1083/jcb.200705182   
375                10.1096/fj.06-6171fje   
376                    10.1104/pp.011019   
377                    10.1104/pp.011019   
378              10.1073/pnas.0501997102   
379              10.1073/pnas.0501997102   

                                        article_citing  
0    <html>\n <body>\n  <pub-id pub-id-type="pmid">...  
1    <html>\n <body>\n  <pub-id pub-id-type="pmid">...  
2    <html>\n <body>\n  <article article-type="rese...  
3    <html>\n <body>\n  <comment>\n   S0891-\n  </c...  
4    <html>\n <body>\n  <article article-type="brie...  
5    <html>\n <body>\n  <pub-id pub-id-type="doi">\...  
6    <html>\n <body>\n  <article article-type="rese...  
7    <html>\n <body>\n  <article article-type="rese...  
8    <html>\n <body>\n  <p>\n   Int_J_Ment_Health_S...  
9    <html>\n <body>\n  <article article-type="rese...  
10   <html>\n <body>\n  <article article-type="data...  
11   <html>\n <body>\n  <article article-type="retr...  
12   <html>\n <body>\n  <p>\n   Sci_Rep/PMC5524909....  
13   <html>\n <body>\n  <pub-id pub-id-type="doi">\...  
14   <html>\n <body>\n  <article article-type="retr...  
15   <html>\n <body>\n  <article-title>\n   Mesench...  
16   <html>\n <body>\n  <article article-type="rese...  
17   <html>\n <body>\n  <p>\n   Biosci_Rep/PMC50258...  
18   <html>\n <body>\n  <article article-type="rese...  
19   <html>\n <body>\n  <year>\n   2016\n  </year>\...  
20   <html>\n <body>\n  <article article-type="rese...  
21   \n   <floats-group>\n    <fig id="ijerph-14

# FIND PAPERS BEFORE AND AFTER RETRACTION, THE RETRACTION NOTE AND THE RETRACTION ITSELF. 
# YOU MAY CONSIDER TO  INTRODUCE A GAP BERORE AND AFTER RETRACTION

In [41]:
from modules import Article_object

list_Articles_citing = []
list_Article_citing_dois = []
list_Article_citing_dates = []
list_Article_citing_references = []


texto = df_articles_citing_retracted_paper['article_citing'].iloc[4]

Article_citing = Article_object.Article_object()
Article_citing.text = texto

Article_citing.get_doi_from_text()
Article_citing.get_date_from_text()
    
Article_citing.get_list_references_dois_and_pmids_from_text()
Article_citing.get_pmid_from_text()
Article_citing.describe()

PMID not found
pmc: 	  
pmid: 	  
doi: 	 10.1128/genomeA.00731-17 
journal: 	  
date: 	 2017-08-10 
text: 	 <html>
 <body>
  <article article-type="brief-report" xmlns:mml="http://www.w3.org/1998/Math/MathML" 
references_dois: 	 


In [42]:
from modules import Article_object

list_Articles_citing = []
list_Article_citing_dois = []
list_Article_citing_pmids = []
list_Article_citing_dates = []
list_Article_citing_references = []


for index, row in df_articles_citing_retracted_paper.iterrows():
    print(index)
    
    #for text in row['article_citing']:
    
    Article_citing = Article_object.Article_object()
    Article_citing.text = row['article_citing']

    Article_citing.get_doi_from_text()
    list_Article_citing_dois.append(Article_citing.doi)

    Article_citing.get_pmid_from_text()
    list_Article_citing_pmids.append(Article_citing.pmid)
    
    Article_citing.get_date_from_text()
    list_Article_citing_dates.append(Article_citing.date)

    Article_citing.get_list_references_dois_and_pmids_from_text()
    list_Article_citing_references.append(Article_citing.references_dois)
    
    

    list_Articles_citing.append(Article_citing)

    #df_Article_ref_retraction = {"doi": row["doi"],
    #                             "pmid":row["pmid"],
    #                             "citing_article": article_ref_retraction,
    #                             "critizising": 1}

    #list_Articles_criticising.append(df_Article_ref_retraction)

"""
df_Articles_citing_retracted_paper = pd.concat([df_articles_citing_retracted_paper, 
                                    pd.DataFrame({"Article": list_Articles_citing,
                                            "date_article_citing": list_Article_citing_dates,
                                            "doi_article_citing": list_Article_citing_dois,
                                            "pmid_article_citing": list_Article_citing_pmids,
                                            "references_in_article_citing": list_Article_citing_references                     
                                            })
                                    ], axis=1)
"""


0
1
2
PMID not found
3
DOI not found
Date not found
4
PMID not found
5
DOI not found
Date not found
6
7
8
DOI not found
9
PMID not found
10
PMID not found
11
12
PMID not found
13
Day or month out of range!!!
day:  
         31
         month:  
         11
         year:  
      2010
     
Date set to: 
day:  29 month:  
         11
         year:  
      2010
     
14
PMID not found
15
DOI not found
Date not found
16
17
DOI not found
Date not found
18
PMID not found
19
DOI not found
Date not found
20
21
PMID not found
22
23
DOI not found
Date not found
24
PMID not found
25
DOI not found
Date not found
26
27
28
29
DOI not found
Date not found
30
DOI not found
Date not found
31
32
PMID not found
33
DOI not found
Date not found
34
DOI not found
Date not found
35
DOI not found
Date not found
36
DOI not found
Date not found
37
DOI not found
Date not found
38
DOI not found
Date not found
39
DOI not found
Date not found
40
DOI not found
Date not found
41
DOI not found
Date not found
42
DOI n

'\ndf_Articles_citing_retracted_paper = pd.concat([df_articles_citing_retracted_paper, \n                                    pd.DataFrame({"Article": list_Articles_citing,\n                                            "date_article_citing": list_Article_citing_dates,\n                                            "doi_article_citing": list_Article_citing_dois,\n                                            "pmid_article_citing": list_Article_citing_pmids,\n                                            "references_in_article_citing": list_Article_citing_references                     \n                                            })\n                                    ], axis=1)\n'

In [43]:
df_Articles_citing_retracted_paper = pd.concat([df_articles_citing_retracted_paper.reset_index(drop=True), 
                                    pd.DataFrame({"Article": list_Articles_citing,
                                            "date_article_citing": list_Article_citing_dates,
                                            "doi_article_citing": list_Article_citing_dois,
                                            "pmid_article_citing": list_Article_citing_pmids,
                                            "pmid_article_citing": list_Article_citing_pmids,
                                            "references_in_article_citing": list_Article_citing_references                     
                                            })
                                    ], axis=1)
df_Articles_citing_retracted_paper

# Let's add the date of the retraction
list_date_retracion = []
for  doi in df_Articles_citing_retracted_paper["DOI_cited"]:
    mask = df_retracted_retraction["doi"] == doi
    list_date_retracion.append(datetime.date(df_retracted_retraction[mask]["date_retraction"].iloc[0]))

df_Articles_citing =  pd.concat([df_Articles_citing_retracted_paper,
                                 pd.DataFrame({"date_doi_cited_retraction": list_date_retracion
                                               })
                                ], axis = 1)

In [44]:
df_Articles_citing.shape

(380, 8)

In [18]:
list_date_retracion[:5]

[datetime.date(2017, 12, 20),
 datetime.date(2017, 12, 20),
 datetime.date(2017, 12, 13),
 datetime.date(2017, 11, 20),
 datetime.date(2017, 11, 16)]

In [45]:
df_Articles_citing.head(5)

DOI_cited  \
0               10.1136/ip.2006.015115   
1               10.1136/ip.2006.015115   
2            10.1186/s40478-016-0393-5   
3  10.1016/j.freeradbiomed.2011.11.027   
4             10.1128/genomeA.00731-17   

                                      article_citing  \
0  <html>\n <body>\n  <pub-id pub-id-type="pmid">...   
1  <html>\n <body>\n  <pub-id pub-id-type="pmid">...   
2  <html>\n <body>\n  <article article-type="rese...   
3  <html>\n <body>\n  <comment>\n   S0891-\n  </c...   
4  <html>\n <body>\n  <article article-type="brie...   

                                             Article date_article_citing  \
0  <modules.Article_object.Article_object object ...          1993-08-30   
1  <modules.Article_object.Article_object object ...          1993-08-30   
2  <modules.Article_object.Article_object object ...          2016-11-18   
3  <modules.Article_object.Article_object object ...                       
4  <modules.Article_object.Article_object object ...          2017-08-10   

           doi_article_citing pmid_article_citing  \
0  10.1136/ip.2006.015115eoc1            25662882   
1  10.1136/ip.2006.015115eoc1            25662882   
2   10.1186/s40478-016-0393-5                       
3                                        22198263   
4    10.1128/genomeA.00731-17                       

  references_in_article_citing date_doi_cited_retraction  
0                                             2017-12-20  
1                                             2017-12-20  
2                                             2017-12-13  
3                                             2017-11-20  
4                                             2017-11-16

In [46]:
df_Articles_citing["Article"].iloc[1].text

'<html>\n <body>\n  <pub-id pub-id-type="pmid">\n   25662882\n  </pub-id>\n  <ref id="pone.0183043.ref009">\n   <label>\n    9\n   </label>\n   <mixed-citation publication-type="journal">\n    <name>\n     <surname>\n      Liisa\n     </surname>\n     <given-names>\n      EHB\n     </given-names>\n    </name>\n    .\n    <article-title>\n     Fatal accidents of older drivers\n    </article-title>\n    .\n    <source>\n     Accident Analysis &amp; Prevention\n    </source>\n    .\n    <year>\n     1993\n    </year>\n    ;\n    <volume>\n     25\n    </volume>\n    (\n    <issue>\n     1\n    </issue>\n    ):\n    <fpage>\n     19\n    </fpage>\n    –\n    <lpage>\n     27\n    </lpage>\n    .\n    <pub-id pub-id-type="pmid">\n     8420531\n    </pub-id>\n   </mixed-citation>\n  </ref>\n  <ref id="pone.0183043.ref010">\n   <label>\n    10\n   </label>\n   <mixed-citation publication-type="journal">\n    <name>\n     <surname>\n      Preusser\n     </surname>\n     <given-names>\n      DF

In [47]:
import numpy as np
df_Articles_citing.replace('', np.nan, inplace=True)
df_Articles_citing2 = df_Articles_citing.dropna(subset = ["date_article_citing"])
df_Articles_citing_clean = df_Articles_citing2.reset_index(drop = True)
df_Articles_citing_clean

DOI_cited  \
0           10.1136/ip.2006.015115   
1           10.1136/ip.2006.015115   
2        10.1186/s40478-016-0393-5   
3         10.1128/genomeA.00731-17   
4            10.3892/mmr.2017.7084   
5        10.1186/s13033-017-0143-5   
6        10.1186/s13033-017-0143-5   
7                10.7150/jca.15753   
8           10.1038/sdata.2016.120   
9               10.1038/ncomms6446   
10              10.1038/ncomms6446   
11              10.1093/nar/gks812   
12           10.3892/mmr.2016.5294   
13              10.1002/cbin.10291   
14    10.1371/journal.pone.0164378   
15    10.1371/journal.pone.0164571   
16    10.1371/journal.pone.0164571   
17         10.1186/1472-6882-14-82   
18            10.1155/2017/6258146   
19             10.1289/ehp.1408282   
20             10.1289/ehp.1408282   
21       10.1186/s12889-017-4082-5   
22    10.1371/journal.pone.0121549   
23    10.1371/journal.pone.0146809   
24      10.1016/j.cell.2013.04.008   
25      10.1016/j.cell.2013.04.008   
26      10.1016/j.cell.2013.04.008   
27      10.1016/j.cell.2013.04.008   
28      10.1016/j.cell.2013.04.008   
29      10.1016/j.cell.2013.04.008   
..                             ...   
176        10.1126/science.1190532   
177        10.1126/science.1190532   
178             10.1101/gad.489409   
179             10.1101/gad.489409   
180        10.1073/pnas.0812470106   
181     10.1016/j.cell.2008.12.021   
182     10.1016/j.cell.2008.12.021   
183  10.1158/0008-5472.CAN-08-0482   
184                 10.1038/nm1491   
185                 10.1038/nm1491   
186                 10.1038/nm1491   
187                 10.1038/nm1491   
188                 10.1038/nm1491   
189                 10.1038/nm1491   
190                 10.1038/nm1491   
191                 10.1038/nm1491   
192       10.1200/JCO.2007.11.0593   
193        10.1073/pnas.0809516106   
194  10.1158/0008-5472.CAN-07-5937   
195        10.1126/science.1124755   
196        10.1126/science.1124755   
197        10.1126/science.1124755   
198  10.1158/0008-5472.CAN-04-4194   
199        10.1073/pnas.0503877102   
200          10.1083/jcb.200705182   
201          10.1083/jcb.200705182   
202          10.1083/jcb.200705182   
203              10.1104/pp.011019   
204        10.1073/pnas.0501997102   
205        10.1073/pnas.0501997102   

                                        article_citing  \
0    <html>\n <body>\n  <pub-id pub-id-type="pmid">...   
1    <html>\n <body>\n  <pub-id pub-id-type="pmid">...   
2    <html>\n <body>\n  <article article-type="rese...   
3    <html>\n <body>\n  <article article-type="brie...   
4    <html>\n <body>\n  <article article-type="rese...   
5    <html>\n <body>\n  <article article-type="rese...   
6    <html>\n <body>\n  <p>\n   Int_J_Ment_Health_S...   
7    <html>\n <body>\n  <article article-type="rese...   
8    <html>\n <body>\n  <article article-type="data...   
9    <html>\n <body>\n  <article article-type="retr...   
10   <html>\n <body>\n  <p>\n   Sci_Rep/PMC5524909....   
11   <html>\n <body>\n  <pub-id pub-id-type="doi">\...   
12   <html>\n <body>\n  <article article-type="retr...   
13   <html>\n <body>\n  <article article-type="rese...   
14   <html>\n <body>\n  <article article-type="rese...   
15   <html>\n <body>\n  <article article-type="rese...   
16   \n   <floats-group>\n    <fig id="ijerph-14-00...   
17   <html>\n <body>\n  <article article-type="rese...   
18   <html>\n <body>\n  <article article-type="retr...   
19   <html>\n <body>\n  <article article-type="rese...   
20   <html>\n <body>\n  <p>\n   Environ_Health_Pers...   
21   <html>\n <body>\n  <article article-type="rese...   
22   \n <p>\n  Biol_Res/PMC4482036.nxml            ...   
23   <html>\n <body>\n  <pub-id pub-id-type="doi">\...   
24   \n              <floats-group>\n              ...   
25   \n                     Lipids_Health_Dis/PMC47...   
26   \n                       Diabetologia/PMC38578...   
27   <html>\n <body>\n  <p>\n   Int_J_Mol_Me

In [48]:
df_Articles_citing_clean.shape

(206, 8)

list_index_is_the_retracted_paper = []

for index, row in df_Articles_citing.iterrows():   
    if row.DOI_cited == row.doi_article_citing:
        list_index_is_the_retracted_paper.append(index)
        #print(row.DOI_cited)
        #print(row.doi_article_citing)
        #print(index)
    
        
df_Retractions_found = df_Articles_citing_retracted_paper_clean.loc[list_index_is_the_retracted_paper]


In [49]:
list_index_is_the_retracted_paper =[]
list_index_is_the_retraction_note = []
for index, row in df_Articles_citing_clean.iterrows():   
    if row.doi_article_citing in list(df_retracted_retraction["doi"]):
        list_index_is_the_retracted_paper.append(index)
        
    if row.doi_article_citing in list(df_retracted_retraction["doi_retraction"]):
        list_index_is_the_retraction_note.append(index)
        #print(row.DOI_cited)
        #print(row.doi_article_citing)
        #print(index)

df_Articles_citing_no_retractions =  df_Articles_citing_clean.drop(list_index_is_the_retracted_paper + 
                                                                   list_index_is_the_retraction_note)


In [50]:
df_Articles_citing_no_retractions.shape

(142, 8)

In [51]:
len(list_index_is_the_retracted_paper)

41

In [52]:
len(list_index_is_the_retraction_note)

24

In [53]:
df_Articles_citing_no_retractions.columns

Index(['DOI_cited', 'article_citing', 'Article', 'date_article_citing',
       'doi_article_citing', 'pmid_article_citing',
       'references_in_article_citing', 'date_doi_cited_retraction'],
      dtype='object')

In [54]:
list_before_retraction = []
list_after_retraction = []

for index, row in df_Articles_citing_no_retractions.iterrows():
    if row['date_article_citing'] < row['date_doi_cited_retraction']:
        list_before_retraction.append(1)
        list_after_retraction.append(0)
    else:
        list_before_retraction.append(0)
        list_after_retraction.append(1)


df_Articles_before_after_retraction = pd.concat([df_Articles_citing_no_retractions.reset_index(drop = True),
                                                pd.DataFrame({"before_retraction": list_before_retraction,
                                                              "after_retraction": list_after_retraction
                                                             })
                                                ], axis = 1)
        

df_Articles_before_after_retraction        


DOI_cited  \
0        10.1186/s13033-017-0143-5   
1               10.1002/cbin.10291   
2     10.1371/journal.pone.0164571   
3       10.1016/j.cell.2013.04.008   
4       10.1016/j.cell.2013.04.008   
5       10.1016/j.cell.2013.04.008   
6       10.1016/j.cell.2013.04.008   
7       10.1016/j.cell.2013.04.008   
8       10.1016/j.cell.2013.04.008   
9           10.1074/jbc.M306295200   
10          10.1074/jbc.M100194200   
11          10.1074/jbc.M100194200   
12      10.1016/j.cell.2010.09.011   
13         10.1186/1742-2094-9-223   
14               10.1038/srep19222   
15               10.1038/srep19222   
16            10.1128/AEM.02433-12   
17            10.1128/AEM.02433-12   
18         10.1126/science.1237242   
19          10.1098/rspb.2009.1774   
20      10.1016/j.cell.2012.02.055   
21      10.1152/ajplung.00134.2011   
22      10.1016/j.cell.2009.03.047   
23      10.1016/j.cell.2009.03.047   
24      10.1016/j.cell.2009.03.047   
25      10.1016/j.cell.2009.03.047   
26      10.1016/j.cell.2009.03.047   
27      10.1016/j.cell.2009.03.047   
28      10.1016/j.cell.2009.03.047   
29      10.1016/j.cell.2009.03.047   
..                             ...   
112        10.1073/pnas.0806674105   
113        10.1073/pnas.0806674105   
114        10.1126/science.1190532   
115        10.1126/science.1190532   
116             10.1101/gad.489409   
117             10.1101/gad.489409   
118        10.1073/pnas.0812470106   
119     10.1016/j.cell.2008.12.021   
120     10.1016/j.cell.2008.12.021   
121  10.1158/0008-5472.CAN-08-0482   
122                 10.1038/nm1491   
123                 10.1038/nm1491   
124                 10.1038/nm1491   
125                 10.1038/nm1491   
126                 10.1038/nm1491   
127                 10.1038/nm1491   
128                 10.1038/nm1491   
129                 10.1038/nm1491   
130       10.1200/JCO.2007.11.0593   
131        10.1073/pnas.0809516106   
132  10.1158/0008-5472.CAN-07-5937   
133        10.1126/science.1124755   
134        10.1126/science.1124755   
135        10.1126/science.1124755   
136  10.1158/0008-5472.CAN-04-4194   
137        10.1073/pnas.0503877102   
138          10.1083/jcb.200705182   
139              10.1104/pp.011019   
140        10.1073/pnas.0501997102   
141        10.1073/pnas.0501997102   

                                        article_citing  \
0    <html>\n <body>\n  <p>\n   Int_J_Ment_Health_S...   
1    <html>\n <body>\n  <article article-type="rese...   
2    <html>\n <body>\n  <article article-type="rese...   
3    \n              <floats-group>\n              ...   
4    \n                     Lipids_Health_Dis/PMC47...   
5    \n                       Diabetologia/PMC38578...   
6    <html>\n <body>\n  <p>\n   Int_J_Mol_Med/PMC53...   
7    \n  Islets/PMC4589993.nxml                    ...   
8    <html>\n <body>\n  <p>\n   J_Clin_Med_Res/PMC5...   
9    \n <p>\n  Lipids_Health_Dis/PMC3544701.nxml   ...   
10   \n <fpage>\n  20436\n </fpage>\n <p>\n  –\n  <...   
11   <html>\n <body>\n  <p>\n   Mol_Cancer/PMC29462...   
12   <html>\n <body>\n  <p>\n   quantile normalizat...   
13   \n  <p>\n   Mol_Autism/PMC3922109.nxml        ...   
14   <html>\n <body>\n  <article article-type="revi...   
15   \n <p>\n  Chin_J_Cancer/PMC5251273.nxml       ...   
16   \n <p>\n  Appl_Microbiol_Biotechnol/PMC5071362...   
17   <html>\n <body>\n  <p>\n   Biotechnol_Biofuels...   
18   <html>\n <body>\n  <article article-type="edit...   
19   \n <p>\n  BMC_Evol_Biol/PMC4587847.nxml       ...   
20   \n <p>\n  BMC_Infect_Dis/PMC4234735.nxml      ...   
21   \n <p>\n  BMC_Pulm_Med/PMC3907130.nxml        ...   
22   <html>\n <body>\n  <article article-type="rese...   
23   <html>\n <body>\n  <p>\n   Arthritis_Res_Ther/...   
24   <html>\n <body>\n  <p>\n   BMC_Cancer/PMC33627...   
25   <html>\n <body>\n  <p>\n   BMC_Cancer/PMC29035...   
26   <html>\n <body>\n  <p>\n   BMC_Cancer/PMC29948...   
27   <html>\n <body>\n  <p>\n   BMC_Genomics

In [55]:
df_Articles_before_after_retraction[df_Articles_before_after_retraction["before_retraction"]==1].shape

(104, 10)

In [30]:
len(list_retraction)

NameError: name 'list_retraction' is not defined

In [56]:
len(df_Articles_before_after_retraction["DOI_cited"].unique())

53

# Summary

61 different DOIs of retracted papers cited by
218 articles.

42 retracted papers, 25 retraction notes, and 152 new manuscripts.

152 new manuscripts:
     45 were published after the retraction
    107 before the retraction.

# Let's analyse the text and find the sentence were the paper was cited

In [57]:
df_Articles_before_after_retraction.tail(1)

DOI_cited  \
141  10.1073/pnas.0501997102   

                                        article_citing  \
141  \n     <floats-group>\n      <fig id="ijms-18-...   

                                               Article date_article_citing  \
141  <modules.Article_object.Article_object object ...          2012-05-04   

          doi_article_citing pmid_article_citing  \
141  10.1186/1465-9921-13-36            11401896   

     references_in_article_citing date_doi_cited_retraction  after_retraction  \
141                           NaN                2009-07-09                 1   

     before_retraction  
141                  0

In [60]:
df_Articles_before_after_retraction.head(3)

DOI_cited  \
0     10.1186/s13033-017-0143-5   
1            10.1002/cbin.10291   
2  10.1371/journal.pone.0164571   

                                      article_citing  \
0  <html>\n <body>\n  <p>\n   Int_J_Ment_Health_S...   
1  <html>\n <body>\n  <article article-type="rese...   
2  <html>\n <body>\n  <article article-type="rese...   

                                             Article date_article_citing  \
0  <modules.Article_object.Article_object object ...          2017-11-30   
1  <modules.Article_object.Article_object object ...          2016-08-16   
2  <modules.Article_object.Article_object object ...          2017-06-05   

       doi_article_citing pmid_article_citing  references_in_article_citing  \
0                     NaN            28503193                           NaN   
1     10.1042/BSR20160174            21980456                           NaN   
2  10.3390/ijerph14060602            24800926                           NaN   

  date_doi_cited_retraction  after_retraction  before_retraction  
0                2017-09-15                 1                  0  
1                2017-07-24                 0                  1  
2                2017-05-26                 1                  0

In [62]:
import imp
import sys
sys.path.insert(0, "./modules")
from modules import get_df_analise_section
#imp.reload()

list_DOI_cited = df_Articles_before_after_retraction.DOI_cited.unique()

df_1sentence = pd.DataFrame() # To get only the sentence citing
df_3sentences = pd.DataFrame() # To get the sentence citing plus the following 3.

for index, row in df_Articles_before_after_retraction.iterrows():
    
    df_analysis_row = get_df_analise_section.get_df(row.DOI_cited, [row.article_citing])
    #df_1sentence = df.append(df_analysis_row)
    df_1sentence = df_1sentence.append(df_analysis_row)

    
    df_3_analysis_row = get_df_analise_section.get_df(row.DOI_cited, [row.article_citing], 3) # We will get 3 additinal sentences
    df_3sentences = df_3sentences.append(df_3_analysis_row)
    
df_1sentence.reset_index(drop = True)
df_3sentences.reset_index(drop = True)

cited_DOI  cited_in_conclusions  cited_in_discussion  \
0        10.1186/s13033-017-0143-5                 False                False   
1               10.1002/cbin.10291                 False                False   
2     10.1371/journal.pone.0164571                 False                 True   
3       10.1016/j.cell.2013.04.008                 False                False   
4       10.1016/j.cell.2013.04.008                 False                False   
5       10.1016/j.cell.2013.04.008                 False                 True   
6       10.1016/j.cell.2013.04.008                 False                False   
7       10.1016/j.cell.2013.04.008                 False                 True   
8       10.1016/j.cell.2013.04.008                 False                False   
9           10.1074/jbc.M306295200                 False                 True   
10          10.1074/jbc.M100194200                 False                False   
11          10.1074/jbc.M100194200                 False                 True   
12      10.1016/j.cell.2010.09.011                 False                False   
13         10.1186/1742-2094-9-223                 False                 True   
14               10.1038/srep19222                 False                False   
15               10.1038/srep19222                 False                False   
16            10.1128/AEM.02433-12                 False                 True   
17            10.1128/AEM.02433-12                 False                False   
18         10.1126/science.1237242                 False                False   
19          10.1098/rspb.2009.1774                 False                 True   
20      10.1016/j.cell.2012.02.055                 False                False   
21      10.1152/ajplung.00134.2011                 False                False   
22      10.1016/j.cell.2009.03.047                 False                False   
23      10.1016/j.cell.2009.03.047                 False                 True   
24      10.1016/j.cell.2009.03.047                 False                False   
25      10.1016/j.cell.2009.03.047                 False                 True   
26      10.1016/j.cell.2009.03.047                 False                 True   
27      10.1016/j.cell.2009.03.047                 False                 True   
28      10.1016/j.cell.2009.03.047                 False                 True   
29      10.1016/j.cell.2009.03.047                 False                 True   
..                             ...                   ...                  ...   
112        10.1073/pnas.0806674105                 False                False   
113        10.1073/pnas.0806674105                 False                False   
114        10.1126/science.1190532                 False                False   
115        10.1126/science.1190532                 False                False   
116             10.1101/gad.489409                 False                 True   
117             10.1101/gad.489409                 False                 True   
118        10.1073/pnas.0812470106                 False                 True   
119     10.1016/j.cell.2008.12.021                 False                False   
120     10.1016/j.cell.2008.12.021                 False                False   
121  10.1158/0008-5472.CAN-08-0482                 False                False   
122                 10.1038/nm1491                 False                False   
123                 10.1038/nm1491                 False                False   
124                 10.1038/nm1491                 False                False   
125                 10.1038/nm1491                 False                False   
126                 10.1038/nm1491                 False                False   
127                 10.1038/nm1491                 False                False   
128                 10.1038/nm1491                 False                False   
129                 10.1038/nm1491                 False

In [329]:
df_1sentence.head(10)

cited_DOI  cited_in_conclusions  cited_in_discussion  \
0     10.1186/s13033-017-0143-5                 False                False   
0            10.1002/cbin.10291                 False                False   
0  10.1371/journal.pone.0164571                 False                 True   
0    10.1016/j.cell.2013.04.008                 False                False   
0    10.1016/j.cell.2013.04.008                 False                False   
0    10.1016/j.cell.2013.04.008                 False                 True   
0    10.1016/j.cell.2013.04.008                 False                False   
0    10.1016/j.cell.2013.04.008                 False                 True   
0    10.1016/j.cell.2013.04.008                 False                False   
0        10.1074/jbc.M306295200                 False                 True   

   cited_in_introduction  cited_in_maintext  \
0                   True               True   
0                  False               True   
0                  False               True   
0                   True               True   
0                  False              False   
0                   True               True   
0                  False               True   
0                   True               True   
0                   True               True   
0                  False              False   

                                          citing_DOI  conclusions_found  \
0                                      DOI not found              False   
0                 \n      10.1042/BSR20160174\n                   False   
0              \n      10.3390/ijerph14060602\n                    True   
0                                      DOI not found               True   
0                                      DOI not found               True   
0  \n                           10.3892/ijmm.2017...              False   
0                                      DOI not found              False   
0                 \n      10.14740/jocmr3114w\n                   False   
0     \n         10.1007/s13238-014-0123-3\n                      False   
0            \n        10.3892/ol.2016.5392\n                     False   

   discussion_found  introduction_found  maintext_found     reference_number  \
0              True                True            True        b1-cpn-15-361   
0              True                True            True                  B49   
0              True                True            True  B34-ijerph-14-00602   
0              True                True            True                   B5   
0             False                True            True                 CR30   
0              True                True            True  b21-ijmm-39-04-0936   
0             False               False            True                   R1   
0              True                True            True                  R04   
0             False                True            True                 CR71   
0              True                True            True      b32-ol-0-0-5392   

  sentence_citing_conclusions  \
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   
0                          []   

                          sentence_citing_discussion  \
0                                                 []   
0                                                 []   
0  [The persistent practice of open defecation ma...   
0                                                 []   
0                                                 []   
0  [Betatrophin is a novel β-cell mitogen reporte...   
0                                                 []   
0  [The reason for the potentially positive assoc...   
0                                                 []   
0  [Another ex

In [349]:
df_1sentence.columns

Index(['cited_DOI', 'cited_in_conclusions', 'cited_in_discussion',
       'cited_in_introduction', 'cited_in_maintext', 'citing_DOI',
       'conclusions_found', 'discussion_found', 'introduction_found',
       'maintext_found', 'reference_number', 'sentence_citing_conclusions',
       'sentence_citing_discussion', 'sentence_citing_intro',
       'sentence_citing_maintext'],
      dtype='object')

In [366]:
len(df_1sentence.cited_DOI.unique())

53

In [351]:
len(df_1sentence)

142

df.sentence_citing_discussion.iloc[0]

# Get Dataframe with preprocessed sentences

In [377]:
#for index, row in df.iterrows():
import re
import numpy as np

"""
df_1sentence[['sentence_citing_conclusions',
       'sentence_citing_discussion', 'sentence_citing_intro',
       'sentence_citing_maintext']]
"""

def remove_citations(text):
    number_citations = text.count("[")
    while number_citations > 0:
        init = text.find("[") 
        end = text.find("]") 
        text = text.replace(text[init : end +1 ], " ") # replace the string between two delimiters
        number_citations -= 1
    return text

def remove_html_tags(text):
    to_clean = re.compile('<.*?>')
    clean_text = re.sub(to_clean, '', text)
    return clean_text
    

def get_df_sentences_preprocessed_and_classified(dataframe):
    
    list_all_citing_citing_sentences = []
    
    for index, row in dataframe.iterrows():

        list_sentences_citing_conclusions = row['sentence_citing_conclusions']
        list_sentences_citing_discussion = row['sentence_citing_discussion']
        list_sentences_citing_intro = row['sentence_citing_intro']
        list_sentences_citing_maintext = row['sentence_citing_maintext']

        if list_sentences_citing_conclusions != None:
            sentences_conclusions = " ".join(list_sentences_citing_conclusions)
            sentences_conclusions = remove_html_tags(sentences_conclusions)
            sentences_conclusions = remove_citations(sentences_conclusions)
        else:
            sentences_conclusions = ""

        if list_sentences_citing_discussion != None:
            sentences_discussion = " ".join(list_sentences_citing_discussion)
            sentences_discussion = remove_html_tags(sentences_discussion)
            sentences_discussion = remove_citations(sentences_discussion)

        else:
            sentences_discussion = ""

        if list_sentences_citing_intro != None:
            sentences_intro = " ".join(list_sentences_citing_intro)
            sentences_intro = remove_html_tags(sentences_intro)
            sentences_intro = remove_citations(sentences_intro)

        else:
            sentences_intro = ""

        if list_sentences_citing_maintext != None:
            sentences_maintext = " ".join(list_sentences_citing_maintext)
            sentences_maintext = remove_html_tags(sentences_maintext)
            sentences_maintext = remove_citations(sentences_maintext)
        else:
            sentences_maintext = ""

        all_sentences = (sentences_maintext + " " + 
                         sentences_discussion + " " +
                         sentences_intro + " " +
                         sentences_maintext)
        if all_sentences != "   ":
            list_all_citing_citing_sentences.append(all_sentences)
        else:
            list_all_citing_citing_sentences.append(np.nan)

    #print(all_sentences)

    dataframe_sentences_classification = pd.concat([df_Articles_before_after_retraction[["DOI_cited", 
                                                                                         "doi_article_citing",
                                                                                         #"pmid_article_citing",
                                                                                         "before_retraction",
                                                                                         "after_retraction"]],
                                            pd.DataFrame({"sentences_citing" : list_all_citing_citing_sentences
                                                          })
                                            ], axis = 1)
    return dataframe_sentences_classification

df1sentence_sentences_classification = get_df_sentences_preprocessed_and_classified(df_1sentence)
df3sentences_sentences_classification = get_df_sentences_preprocessed_and_classified(df_3sentences)

In [378]:
len(df1sentence_sentences_classification)

142

In [379]:
df_Articles_before_after_retraction.head(1)

DOI_cited  \
0  10.1186/s13033-017-0143-5   

                                      article_citing  \
0  <html>\n <body>\n  <p>\n   Int_J_Ment_Health_S...   

                                             Article date_article_citing  \
0  <modules.Article_object.Article_object object ...          2017-11-30   

  doi_article_citing pmid_article_citing  references_in_article_citing  \
0                NaN            28503193                           NaN   

  date_doi_cited_retraction  after_retraction  before_retraction  
0                2017-09-15                 1                  0

In [380]:
df1sentence_sentences_classification.tail(3)

DOI_cited        doi_article_citing  before_retraction  \
139        10.1104/pp.011019  10.1186/1472-6882-13-155                  0   
140  10.1073/pnas.0501997102      10.3390/ijms18122636                  1   
141  10.1073/pnas.0501997102   10.1186/1465-9921-13-36                  0   

     after_retraction                                   sentences_citing  
139                 1  For example, flavonoids showed antimicrobial a...  
140                 0  Whereas cardiodynamic (heart rate   reduction,...  
141                 1  Others have reported that changes in occludin ...

In [381]:
#df.dropna(subset = "sentences_citing")

df1sentence_sentences_classification  = df1sentence_sentences_classification.dropna(subset = ["sentences_citing"])

df3sentences_sentences_classification  = df3sentences_sentences_classification.dropna(subset = ["sentences_citing"])

In [399]:
len(df1sentence_sentences_classification.doi_article_citing.unique())

111

In [395]:
df1sentence_sentences_classification.head(3)


DOI_cited      doi_article_citing  before_retraction  \
0     10.1186/s13033-017-0143-5                     NaN                  0   
1            10.1002/cbin.10291     10.1042/BSR20160174                  1   
2  10.1371/journal.pone.0164571  10.3390/ijerph14060602                  0   

   after_retraction                                   sentences_citing  
0                 1   1)  Although there has been meaningful debate...  
1                 0  To verify the activation of FOXO1 in our model...  
2                 1  Boiling or chlorinating water at the point of ...

In [389]:
len(df1sentence_sentences_classification.DOI_cited.unique())

47

In [407]:
df_sentences_before = df1sentence_sentences_classification[df1sentence_sentences_classification["before_retraction"] == 1]
df_sentences_after = df1sentence_sentences_classification[df1sentence_sentences_classification["after_retraction"] == 1]
print("df_sentences_before: ", df_sentences_before.shape)
print("df_sentences_after: ", df_sentences_after.shape)

df_3sentences_before = df3sentences_sentences_classification[df3sentences_sentences_classification["before_retraction"] == 1]
df_3sentences_after = df3sentences_sentences_classification[df3sentences_sentences_classification["after_retraction"] == 1]


df_sentences_before:  (90, 5)
df_sentences_after:  (32, 5)


In [456]:
len(df_sentences_after.DOI_cited.unique())

24

In [474]:
#####################################################################################

# Let's analyse the documents

* How many the documents from df_sentences_before cite DOIs that are not cited after retraction? 
* How many cite DOIs that are also cited after retraction?

* How many the documents from df_sentences_after cite DOIs that are not cited before retraction? 
* How many cite DOIs that are also cited before retraction?


In [424]:
"""
df_sentences_before_selected_DOIs = df_sentences_before[df_sentences_before["DOI_cited"].isin(df_sentences_after.DOI_cited.unique())]

len(df_sentences_before_selected_DOIs.DOI_cited.unique())

len(df_sentences_before_selected_DOIs.doi_article_citing.unique())
"""

35

In [453]:
dois_cited_before = set(df_sentences_before["DOI_cited"])
dois_cited_after = set(df_sentences_after["DOI_cited"])

doi_cited_before_and_after = dois_before.intersection(dois_cited_after)
doi_cited_before_and_after 

doi_cited_only_before = dois_cited_before.difference(dois_cited_after)
doi_cited_only_before

doi_cited_only_after = dois_after.difference(dois_cited_before)
doi_cited_only_after

print("\n unique dois_cited_before ", len(dois_cited_before))
print("\n unique dois_cited_after ", len(dois_cited_after))
print("\n unique doi_cited_before_and_after ", len(doi_cited_before_and_after))
print("\n unqiue doi_cited_only_before ", len(doi_cited_only_before))
print("\n unique doi_cited_only_after ", len(doi_cited_only_after))




 unique dois_cited_before  35

 unique dois_cited_after  24

 unique doi_cited_before_and_after  12

 unqiue doi_cited_only_before  23

 unique doi_cited_only_after  12


In [480]:
print("\n number documents citing after - DOIS that are only cited after: ")
df_sentences_after_only = df_sentences_after[df_sentences_after.DOI_cited.isin(doi_cited_only_after)]
print(df_sentences_after_only_after.shape[0])

print("\n number documents citing after - DOIS that are cited before and after:")
df_sentences_after_not_only = df_sentences_after[df_sentences_after.DOI_cited.isin(doi_cited_before_and_after)]
print(df_sentences_after_not_only.shape[0])

print("\n number documents citing before - DOIS that are only cited before: ")
df_sentences_before_only = df_sentences_before[df_sentences_before.DOI_cited.isin(doi_cited_only_before)]
print(df_sentences_before_only.shape[0])

print("\n number documents citing before - DOIS that are cited before and after:")
df_sentences_before_not_only = df_sentences_before[df_sentences_before.DOI_cited.isin(doi_cited_before_and_after)]
print(df_sentences_before_not_only.shape[0])



 number documents citing after - DOIS that are only cited after: 
15

 number documents citing after - DOIS that are cited before and after:
17

 number documents citing before - DOIS that are only cited before: 
53

 number documents citing before - DOIS that are cited before and after:
37


In [477]:
df_sentences_before.shape

(90, 5)

# Summary 3:

90 documents before retractions:
    
    - 53 documents cite DOIs (23 unique DOIs) that are not cited after retraction.
    - 37 documents cite DOIs (12 unique DOIs) that are also cited after retraction.

    - 83 unique doi_article_citing - len(df_sentences_before.doi_article_citing.unique())
        There are 7 documents without the DOIS_citing     
    - 35 unique DOI_cited - len(df_sentences_before.DOI_cited.unique())
        Of those 23 DOI are only cited before. 12 before and after.


32 documents after retraction:
    
    - 15 documents cite DOIs (12 unique DOIs) that are not cited before retraction.
    - 17 documents cite DOIs (12 unique DOIs) that are also cited before retraction.


    - 29 unique doi_article_citing - len(df_sentences_after.doi_article_citing.unique())
        There are 3 documents without the DOIS_citing        
    - 24 unique DOI_cited - len(df_sentences_after.DOI_cited.unique())
        Of those 12 DOI are only cited after. 12 before and after.
    

We will select documents before retraction only if they also have been cited after retraction.


#####################################################################################

# Summary 2 

61 different DOIs of retracted papers cited by
218 articles.

42 retracted papers, 25 retraction notes, and 152 new manuscripts.

152 new manuscripts:
     45 were published after the retraction
    107 before the retraction.
    
* The algorithm is able to find the text citing in:
    
127 manuscripts:
    36 were published after retraction
    91 were published before retraction


# USING NLTK

# I will take here 25 entries for before and after for training and 11 for test

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer() # Checking all_words later, I think it doesn't work properly

from nltk.stem import PorterStemmer
ps = PorterStemmer()

sentence = "I will include: include, including and included. in my list of included words."

word_tokens = word_tokenize(sentence)

filt_sentence = []
list_words = []

for word in word_tokens:
    #filt_sentence.append(lemmatizer.lemmatize(word.lower()))    
    #list_words.append(lemmatizer.lemmatize(word.lower()))
    filt_sentence.append(ps.stem(word.lower()))    
    list_words.append(ps.stem(word.lower()))            
    

print(filt_sentence)
print(list_words)
set(list_words)

# I will take here 25 entries for before and after for training and 11 for test
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer() # Checking all_words later, I think it doesn't work properly

from nltk.stem import PorterStemmer
ps = PorterStemmer()

# We will keep words from stopwords that could be used when criticing
stop_words = set(stopwords.words('english')) - set(["while", "against", "before", "after", "just", "don", 
                                                    "don't", "should", "should've", "no", "nor", "not",
                                                    "ain", "aren", "aren't", "couldn", "couldn't", "didn",
                                                    "didn't", "doesn", "doesn't", "hadn", "hadn't", "hasn",
                                                    "hasn't", "haven", "haven't", "isn", "isn't", "ma",
                                                    "mightn", "mightn't", "mustn", "mustn't", "needn", 
                                                    "needn't", "shan", "shan't", "shouldn", "shouldn't",
                                                    "wasn", "wasn't", "weren", "weren't", "won"," won't",
                                                    "wouldn", "wouldn't"])

# We must create a list of tuplas [(tokenized_sentence, type_of_sentence)]

###list_all_words = [] # I will use later to count word frequencies



list_all_words = [] 
document_before = []
for sentence in df_sentences_before.sentences_citing:
    word_tokens = word_tokenize(sentence)
    filtered_sentence = []
    for word in word_tokens:
        if (len(word) > 2) and (not word in stop_words):
            #print(word)
            #filtered_sentence.append(lemmatizer.lemmatize(word.lower()))    
            #list_all_words.append(lemmatizer.lemmatize(word.lower()))
            filtered_sentence.append(ps.stem(word.lower()))    
            list_all_words.append(ps.stem(word.lower()))            
    document_before.append((filtered_sentence, "before"))

document_after = []
for sentence in df_sentences_after.sentences_citing:
    word_tokens = word_tokenize(sentence)
    filtered_sentence = []
    for word in word_tokens:
        if (len(word) > 2) and (not word in stop_words):
            #print(word)
            #filtered_sentence.append(lemmatizer.lemmatize(word.lower()))
            #list_all_words.append(lemmatizer.lemmatize(word.lower()))
            filtered_sentence.append(ps.stem(word.lower()))
            list_all_words.append(ps.stem(word.lower()))            
    document_after.append((filtered_sentence, "after"))
    

# documents = document_before[0:37] + document_after[0:37]  # I will create documents later


document_before_training = document_before[:25]
document_before_test = document_before[25:37]  

document_after_training = document_after[:25]
document_after_test = document_after[25:37]  


import random
random.seed(21)

documents_training = document_before_training + document_after_training
random.shuffle(document_training)

documents_test = document_before_test + document_after_test
random.shuffle(documents_test)                                  

# All the documents I have (50% before and 50% after)
documents = documents_training + documents_test
random.shuffle(documents)

# All the documents
documents_all = document_before + document_after
random.shuffle(documents_all)




"""
for word in word_tokenize(sentence) 
    document_before.append((word_tokenize(sentence), "before"))

document_after = []
for sentence in df_sentences_after.sentences_citing:
    document_after.append((word_tokenize(sentence), "after"))


number_training_entries = 25
number_test_entries = 11


df_sentences_before = df_sentences_before.reset_index(drop = True)
df_sentences_after = df_sentences_after.reset_index(drop = True)

df_sentences_before_training = df_sentences_before.loc[range(25)]  
df_sentences_before_test = df_sentences_before.loc[range(25, 37)]  

df_sentences_after_training = df_sentences_after.loc[range(25)]  
df_sentences_after_test = df_sentences_after.loc[range(25, 36)]  



""" 




In [238]:
import re
s = "3234234, ke234, 234234k3, 2342kli42, 425, 245lk34kl2, /323, -32342, --3234, \\, '876' 'hjgkk'"
# s = re.sub("[\d?,-?,+?,/?,'?]", "", s) 
s = re.sub("[\d?, '\W+]", "", s) 

#s = re.sub('\W+',' ', s)
s

'kekklilkklhjgkk'

In [406]:
df_sentences_before.head(3)

DOI_cited   doi_article_citing  before_retraction  \
1          10.1002/cbin.10291  10.1042/BSR20160174                  1   
3  10.1016/j.cell.2013.04.008                  NaN                  1   
6  10.1016/j.cell.2013.04.008                  NaN                  1   

   after_retraction                                   sentences_citing  
1                 0  To verify the activation of FOXO1 in our model...  
3                 0  in a S961-induced insulin resistance mouse mod...  
6                 0   14  ,  15  These data suggest the existence o...

In [402]:

import re
import random

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer() # Checking all_words later, I think it doesn't work properly

from nltk.stem import PorterStemmer
ps = PorterStemmer()

# We will keep words from stopwords that could be used when criticing
stop_words = set(stopwords.words('english')) - set(["while", "against", "before", "after", "just", "don", 
                                                    "don't", "should", "should've", "no", "nor", "not",
                                                    "ain", "aren", "aren't", "couldn", "couldn't", "didn",
                                                    "didn't", "doesn", "doesn't", "hadn", "hadn't", "hasn",
                                                    "hasn't", "haven", "haven't", "isn", "isn't", "ma",
                                                    "mightn", "mightn't", "mustn", "mustn't", "needn", 
                                                    "needn't", "shan", "shan't", "shouldn", "shouldn't",
                                                    "wasn", "wasn't", "weren", "weren't", "won"," won't",
                                                    "wouldn", "wouldn't"])

def get_stemmed_sentences_and_list_words(list_sentences, classification):
    list_all_words = []
    document_stemmed = []
    for sentence in list_sentences:
        filtered_sentence = []
        word_tokens = word_tokenize(sentence)
        for word in word_tokens:
            word = re.sub("[\d?, '\W+]", "", word) #Let's remove any number
            if (len(word) > 2) and (not word in stop_words):
                filtered_sentence.append(ps.stem(word.lower()))    
                list_all_words.append(ps.stem(word.lower()))            
        document_stemmed.append((filtered_sentence, classification))
    
    return (document_stemmed, list_all_words)


# I will take here 25 entries for before and after for training and 11 for test

# for df_sentences_before

document_before = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "before")[0]
random.shuffle(document_before)
list_all_words_before = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "before")[1]

document_after = get_stemmed_sentences_and_list_words(df_sentences_after.sentences_citing, "after")[0]
random.shuffle(document_after)
list_all_words_after = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "after")[1]

list_all_words = sorted(list(set(list_all_words_before + list_all_words_after)))
 

#for df_3sentences_before

document_3_before = get_stemmed_sentences_and_list_words(df_3sentences_before.sentences_citing, "before")[0]
random.shuffle(document_3_before)
list_3_all_words_before = get_stemmed_sentences_and_list_words(df_3sentences_before.sentences_citing, "before")[1]

document_3_after = get_stemmed_sentences_and_list_words(df_3sentences_after.sentences_citing, "after")[0]
random.shuffle(document_3_after)
list_3_all_words_after = get_stemmed_sentences_and_list_words(df_3sentences_after.sentences_citing, "after")[1]

list_3_all_words = sorted(list(set(list_3_all_words_before + list_3_all_words_after)))
    
    
    

In [503]:
# Let's get all the papers from df_sentences_after_only (X), and from df_sentences_after_not_only (Y)
# and X papers from df_sentences_before_only, and Y documents from df_sentences_before_and_after. 
# When selecting the documents for df_sentences_before_and_after we must be sure that we are not selecting 
# all the papers the same doi (for each cited doi, we want to have the number of douments citing before and 
# after)

df_sentences_before_only_selection = df_sentences_before_only.sample(frac=1) #https://stackoverflow.com/questions/29576430/shuffle-dataframe-rows
df_sentences_before_only_selection = df_sentences_before_only_selection.drop_duplicates(subset = "DOI_cited")
df_sentences_before_only_selection = df_sentences_before_only_selection[:df_sentences_after_only.shape[0]]
df_sentences_before_only_selection

list_doi_added = []
for doi in list(df_sentences_after_not_only.DOI_cited):
    print(doi)
    times_doi_in_list = list(df_sentences_after_not_only.DOI_cited).count(doi)
    print(times_doi_in_list)
    print("times_doi in df_sentences_after_not_only: ", df_sentences_after_not_only[df_sentences_after_not_only["DOI_cited"] == doi].shape[0])
    print("times_doi in df_sentences_before_not_only: ", df_sentences_before_not_only[df_sentences_before_not_only["DOI_cited"] == doi].shape[0])
    
   

10.1016/j.cell.2013.04.008
2
times_doi in df_sentences_after_not_only:  2
times_doi in df_sentences_before_not_only:  3
10.1016/j.cell.2013.04.008
2
times_doi in df_sentences_after_not_only:  2
times_doi in df_sentences_before_not_only:  3
10.1016/j.cell.2009.03.047
2
times_doi in df_sentences_after_not_only:  2
times_doi in df_sentences_before_not_only:  17
10.1016/j.cell.2009.03.047
2
times_doi in df_sentences_after_not_only:  2
times_doi in df_sentences_before_not_only:  17
10.1038/ni.1741
1
times_doi in df_sentences_after_not_only:  1
times_doi in df_sentences_before_not_only:  3
10.1083/jcb.200702053
1
times_doi in df_sentences_after_not_only:  1
times_doi in df_sentences_before_not_only:  1
10.1016/j.molcel.2009.09.031
1
times_doi in df_sentences_after_not_only:  1
times_doi in df_sentences_before_not_only:  1
10.1038/nature09301
1
times_doi in df_sentences_after_not_only:  1
times_doi in df_sentences_before_not_only:  3
10.1093/jmcb/mjq011
2
times_doi in df_sentences_after_not_o

In [ ]:
#

df_


document_before = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "before")[0]
random.shuffle(document_before)
list_all_words_before = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "before")[1]

document_after = get_stemmed_sentences_and_list_words(df_sentences_after.sentences_citing, "after")[0]
random.shuffle(document_after)
list_all_words_after = get_stemmed_sentences_and_list_words(df_sentences_before.sentences_citing, "after")[1]

list_all_words = sorted(list(set(list_all_words_before + list_all_words_after)))
 

#for df_3sentences_before

document_3_before = get_stemmed_sentences_and_list_words(df_3sentences_before.sentences_citing, "before")[0]
random.shuffle(document_3_before)
list_3_all_words_before = get_stemmed_sentences_and_list_words(df_3sentences_before.sentences_citing, "before")[1]

document_3_after = get_stemmed_sentences_and_list_words(df_3sentences_after.sentences_citing, "after")[0]
random.shuffle(document_3_after)
list_3_all_words_after = get_stemmed_sentences_and_list_words(df_3sentences_after.sentences_citing, "after")[1]

list_3_all_words = sorted(list(set(list_3_all_words_before + list_3_all_words_after)))
 

In [403]:
document_after

[(['clinic',
   'risk',
   'msc',
   'due',
   'use',
   'fb',
   'becom',
   'recogn',
   'cultur',
   'media',
   'contain',
   'varieti',
   'growth',
   'factor',
   'could',
   'replac',
   'anim',
   'serum',
   'humanserum',
   'platelet',
   'lysat',
   'deriv',
   'sphingosinephosph',
   'develop',
   'clinic',
   'risk',
   'msc',
   'due',
   'use',
   'fb',
   'becom',
   'recogn',
   'cultur',
   'media',
   'contain',
   'varieti',
   'growth',
   'factor',
   'could',
   'replac',
   'anim',
   'serum',
   'humanserum',
   'platelet',
   'lysat',
   'deriv',
   'sphingosinephosph',
   'develop',
   'clinic',
   'risk',
   'msc',
   'due',
   'use',
   'fb',
   'becom',
   'recogn',
   'cultur',
   'media',
   'contain',
   'varieti',
   'growth',
   'factor',
   'could',
   'replac',
   'anim',
   'serum',
   'humanserum',
   'platelet',
   'lysat',
   'deriv',
   'sphingosinephosph',
   'develop'],
  'after'),
 (['boil',
   'chlorin',
   'water',
   'point',
   'use',
 

In [404]:
len(documents_test)

20

In [334]:
"""# I will take here 25 entries for before and after for training and 11 for test
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer() # Checking all_words later, I think it doesn't work properly

from nltk.stem import PorterStemmer
ps = PorterStemmer()
    
######################"""


print("Number documents\n - before retraction: ", len(document_before),
     "\n - afer retration: ", len(document_after))

ratio_test = 0.32
number_documents_test = round(len(document_after) * ratio_test)
number_documents_training = len(document_after) - number_documents_test

print("\nSpliting data (after retraction):\n - training: ", number_documents_training,
     "\n - test ", number_documents_test)

#####################


document_before_training = document_before[:number_documents_training]
document_before_test = document_before[number_documents_training:(number_documents_training + 
                                                                  number_documents_test)]  

document_after_training = document_after[:number_documents_training]
document_after_test = document_after[number_documents_training:(number_documents_training + 
                                                                  number_documents_test)]  


import random
random.seed(21)

documents_training = document_before_training + document_after_training
random.shuffle(documents_training)

documents_test = document_before_test + document_after_test
random.shuffle(documents_test)                                  

# All the documents I have (50% before and 50% after)
documents = documents_training + documents_test
random.shuffle(documents)

# All the documents
documents_all = document_before + document_after
random.shuffle(documents_all)


######################


"""document_3_before_training = document_3_before[:25]
document_3_before_test = document_3_before[25:37]  

document_3_after_training = document_3_after[:25]
document_3_after_test = document_3_after[25:37]  
"""
document_3_before_training = document_before[:number_documents_training]
document_3_before_test = document_before[number_documents_training:(number_documents_training + 
                                                                  number_documents_test)]  

document_3_after_training = document_3_after[:number_documents_training]
document_3_after_test = document_3_after[number_documents_training:(number_documents_training + 
                                                                  number_documents_test)]  


import random
random.seed(21)

documents_3_training = document_3_before_training + document_3_after_training
random.shuffle(documents_3_training)

documents_3_test = document_3_before_test + document_3_after_test
random.shuffle(documents_3_test)                                  

# All the documents I have (50% before and 50% after)
documents_3 = documents_3_training + documents_3_test
random.shuffle(documents_3)

# All the documents
documents_3_all = document_3_before + document_3_after
random.shuffle(documents_3_all)


Number documents
 - before retraction:  90 
 - afer retration:  32

Spliting data (after retraction):
 - training:  22 
 - test  10


In [335]:
#random.seed(21)
#random.shuffle(documents_test)
documents_test[10]
#len(documents_test)

for (text, clasi) in documents_test:
    print(clasi)
len(documents_test)

before
after
before
after
before
before
after
before
after
after
after
after
before
after
before
before
before
before
after
after


20

In [336]:
len(documents_3_training)

44

# Word frequencies

In [337]:
len(all_words)

893

In [338]:
import nltk 

all_words = sorted(list(set(list_all_words)))
all_words

all_words_freq = nltk.FreqDist(all_words) 
#FreqDist gives the frequency of each vocabulary item in the text

word_features = list(all_words_freq.keys())  #[:3000] # add if you want to limit the number of features

word_features # Those are all teh words in the document. Each word is a feature

all_words_freq

#########################3

all_words_3 = sorted(list(set(list_3_all_words)))
all_words_3

all_words_3_freq = nltk.FreqDist(all_words_3) 
#FreqDist gives the frequency of each vocabulary item in the text

word_features_3 = list(all_words_3_freq.keys())  #[:3000] # add if you want to limit the number of features

word_features_3 # Those are all teh words in the document. Each word is a feature

all_words_3_freq


FreqDist({'abbrevi': 1,
          'aberr': 1,
          'abil': 1,
          'abl': 1,
          'ablat': 1,
          'abnorm': 1,
          'absenc': 1,
          'absent': 1,
          'abstract': 1,
          'abund': 1,
          'academi': 1,
          'acceler': 1,
          'accept': 1,
          'access': 1,
          'accord': 1,
          'account': 1,
          'accumben': 1,
          'accumul': 1,
          'accuraci': 1,
          'achiev': 1,
          'acid': 1,
          'across': 1,
          'act': 1,
          'actin': 1,
          'actinrich': 1,
          'action': 1,
          'activ': 1,
          'activationinduc': 1,
          'acut': 1,
          'ad': 1,
          'adapt': 1,
          'addit': 1,
          'adenomat': 1,
          'adhes': 1,
          'adipocyt': 1,
          'adiponectin': 1,
          'adipos': 1,
          'adiposederiv': 1,
          'adjac': 1,
          'administ': 1,
          'administr': 1,
          'admsc': 1,
          'admscs

In [339]:
def find_features(document):
    words = set(document)
    features = {}
    for w in word_features:
        features[w] = (w in words)

    return features

featuresets = [(find_features(article), cat) for (article, cat) in documents]
featuresets_all = [(find_features(article), cat) for (article, cat) in documents_all] 

featuresets_3 = [(find_features(article), cat) for (article, cat) in documents_3]
featuresets_3_all = [(find_features(article), cat) for (article, cat) in documents_3_all] 


In [340]:
featuresets_all[2]

({'abil': False,
  'abl': False,
  'abnorm': False,
  'absenc': False,
  'abstract': False,
  'academi': False,
  'accumul': False,
  'achiev': False,
  'act': False,
  'actin': False,
  'activ': False,
  'ad': False,
  'adapt': False,
  'addit': False,
  'adipocyt': False,
  'adiponectin': False,
  'adipos': False,
  'adult': False,
  'affect': False,
  'after': False,
  'age': False,
  'aggreg': False,
  'aggress': False,
  'ago': False,
  'agon': False,
  'akt': False,
  'albeit': False,
  'alia': False,
  'allel': False,
  'allow': False,
  'alon': False,
  'also': False,
  'although': False,
  'america': False,
  'among': False,
  'amphibian': False,
  'ampknul': False,
  'amplifi': False,
  'andor': False,
  'anesthet': False,
  'aneuploidi': False,
  'angiogenesi': False,
  'angptl': False,
  'annot': False,
  'anoiki': False,
  'anoth': False,
  'antagonist': False,
  'antiax': False,
  'antibacteri': False,
  'antibodi': False,
  'antibodyproduc': False,
  'antifung': False,
 

In [341]:
len(featuresets)

64

In [342]:
len(featuresets_3)

64

In [343]:
featuresets_3[4]

({'abil': False,
  'abl': False,
  'abnorm': False,
  'absenc': False,
  'abstract': False,
  'academi': False,
  'accumul': False,
  'achiev': False,
  'act': False,
  'actin': False,
  'activ': True,
  'ad': False,
  'adapt': False,
  'addit': False,
  'adipocyt': False,
  'adiponectin': False,
  'adipos': False,
  'adult': False,
  'affect': False,
  'after': False,
  'age': False,
  'aggreg': False,
  'aggress': False,
  'ago': False,
  'agon': False,
  'akt': False,
  'albeit': False,
  'alia': False,
  'allel': False,
  'allow': False,
  'alon': False,
  'also': False,
  'although': False,
  'america': False,
  'among': False,
  'amphibian': False,
  'ampknul': False,
  'amplifi': False,
  'andor': False,
  'anesthet': False,
  'aneuploidi': False,
  'angiogenesi': False,
  'angptl': False,
  'annot': False,
  'anoiki': False,
  'anoth': False,
  'antagonist': False,
  'antiax': False,
  'antibacteri': False,
  'antibodi': False,
  'antibodyproduc': False,
  'antifung': False,
  

In [344]:
featuresets[1]

({'abil': False,
  'abl': False,
  'abnorm': False,
  'absenc': False,
  'abstract': False,
  'academi': False,
  'accumul': False,
  'achiev': False,
  'act': False,
  'actin': False,
  'activ': False,
  'ad': False,
  'adapt': False,
  'addit': False,
  'adipocyt': False,
  'adiponectin': False,
  'adipos': False,
  'adult': False,
  'affect': False,
  'after': False,
  'age': False,
  'aggreg': False,
  'aggress': False,
  'ago': False,
  'agon': False,
  'akt': False,
  'albeit': False,
  'alia': False,
  'allel': False,
  'allow': False,
  'alon': False,
  'also': False,
  'although': False,
  'america': False,
  'among': False,
  'amphibian': False,
  'ampknul': False,
  'amplifi': False,
  'andor': False,
  'anesthet': False,
  'aneuploidi': False,
  'angiogenesi': False,
  'angptl': False,
  'annot': False,
  'anoiki': False,
  'anoth': False,
  'antagonist': False,
  'antiax': False,
  'antibacteri': False,
  'antibodi': False,
  'antibodyproduc': False,
  'antifung': False,
 

In [345]:
# set that we'll train our classifier with
training_set = featuresets[:320]

# set that we'll test against.
testing_set = featuresets[20:]

classifier = nltk.NaiveBayesClassifier.train(training_set)
print("Classifier accuracy percent:",(nltk.classify.accuracy(classifier, testing_set))*100)

##################################

# set that we'll train our classifier with
training_set_3 = featuresets_3[:20]

# set that we'll test against.
testing_set_3 = featuresets_3[20:]

classifier_3 = nltk.NaiveBayesClassifier.train(training_set_3)
print("Classifier accuracy percent 3 sentences:",(nltk.classify.accuracy(classifier_3, testing_set_3))*100)




Classifier accuracy percent: 90.9090909090909
Classifier accuracy percent 3 sentences: 59.09090909090909


In [347]:
classifier_3.show_most_informative_features(40)

Most Informative Features
                  recent = True            after : before =      3.4 : 1.0
                 express = True            after : before =      2.6 : 1.0
                  differ = True            after : before =      2.4 : 1.0
                 possibl = True            after : before =      2.4 : 1.0
                approach = True            after : before =      2.4 : 1.0
              differenti = True            after : before =      2.4 : 1.0
                   model = True            after : before =      2.4 : 1.0
                     set = True            after : before =      2.4 : 1.0
                  effect = True            after : before =      2.4 : 1.0
                 respons = True            after : before =      2.4 : 1.0
                   while = True            after : before =      2.4 : 1.0
                function = True            after : before =      2.4 : 1.0
                   regul = True            after : before =      2.4 : 1.0

In [96]:
# set that we'll train our classifier with
training_set_all = featuresets_all[:50]

# set that we'll test against.
testing_set_all = featuresets_all[50:70]

classifier_all = nltk.NaiveBayesClassifier.train(training_set_all)
print("Classifier accuracy percent:",(nltk.classify.accuracy(classifier, testing_set_all))*100)

Classifier accuracy percent: 100.0


In [97]:
classifier_all.show_most_informative_features(25)

Most Informative Features


# Preprocessing data with word_frequency_sentence_technicisms module

In [ ]:
#for index, row in df.iterrows():
import re
import numpy as np
from modules import words_frec_analysis_get_sentence

"""
df[['sentence_citing_conclusions',
       'sentence_citing_discussion', 'sentence_citing_intro',
       'sentence_citing_maintext']]
"""

def remove_citations(text):
    number_citations = text.count("[")
    while number_citations > 0:
        init = text.find("[") 
        end = text.find("]") 
        text = text.replace(text[init : end +1 ], " ") # replace the string between two delimiters
        number_citations -= 1
    return text

def remove_html_tags(text):
    to_clean = re.compile('<.*?>')
    clean_text = re.sub(to_clean, '', text)
    return clean_text
    
    
list_all_citing_citing_sentences = []

for index, row in df.iterrows():
    list_sentences_citing_conclusions = row['sentence_citing_conclusions']
    list_sentences_citing_discussion = row['sentence_citing_discussion']
    list_sentences_citing_intro = row['sentence_citing_intro']
    list_sentences_citing_maintext = row['sentence_citing_maintext']
    
    if list_sentences_citing_conclusions != None:
        sentences_conclusions = " ".join(list_sentences_citing_conclusions)
        #sentences_conclusions = remove_html_tags(sentences_conclusions)
        #sentences_conclusions = remove_citations(sentences_conclusions)
        sentences_conclusions = words_frec_analysis_get_sentence.filter_sentence(sentences_conclusions)
    else:
        sentences_conclusions = ""
        
    if list_sentences_citing_discussion != None:
        sentences_discussion = " ".join(list_sentences_citing_discussion)
        #sentences_discussion = remove_html_tags(sentences_discussion)
        #sentences_discussion = remove_citations(sentences_discussion)
        sentences_discussion = words_frec_analysis_get_sentence.filter_sentence(sentences_discussion)


    else:
        sentences_discussion = ""
    
    if list_sentences_citing_intro != None:
        sentences_intro = " ".join(list_sentences_citing_intro)
        #sentences_intro = remove_html_tags(sentences_intro)
        #sentences_intro = remove_citations(sentences_intro)
        sentences_intro = words_frec_analysis_get_sentence.filter_sentence(sentences_intro)
        
    else:
        sentences_intro = ""

    if list_sentences_citing_maintext != None:
        sentences_maintext = " ".join(list_sentences_citing_maintext)
        #sentences_maintext = remove_html_tags(sentences_maintext)
        #sentences_maintext = remove_citations(sentences_maintext)
        sentences_maintext = words_frec_analysis_get_sentence.filter_sentence(sentences_maintext)

    else:
        sentences_maintext = ""

    all_sentences = (sentences_maintext + " " + 
                     sentences_discussion + " " +
                     sentences_intro + " " +
                     sentences_maintext)
    if all_sentences != "   ":
        list_all_citing_citing_sentences.append(all_sentences)
    else:
        list_all_citing_citing_sentences.append(np.nan)

#print(all_sentences)

df_sentences_classification2 = pd.concat([df_Articles_before_after_retraction[["before_retraction",
                                                                             "after_retraction"]],
                                        pd.DataFrame({"sentences_citing" : list_all_citing_citing_sentences
                                                      })
                                        ], axis = 1)
df_sentences_classification2

In [ ]:
words_frec_analysis_get_sentence.analysis(list(df_sentences_classification2.sentences_citing))

# Let's get more sentences than just the one citing

Sentence citing is obtained using get_df_analise_section.get_df

I makes use of find_sention.sentences(paragraph, ref_id). 

I have modified def sentences(text, ref_id). 
Now def sentences(text,ref_id, Number_of additional_sentences_after_citing_sentence), and therefore:
def get_df(DOI, list_papers, number_of_additional_sentences_after_citation)



In [ ]:
import imp
import sys
sys.path.insert(0, "./modules")
from modules import get_df_analise_section
imp.reload(get_df_analise_section)

list_DOI_cited = df_Articles_before_after_retraction.DOI_cited.unique()

df_3_sentences = pd.DataFrame()
for index, row in df_Articles_before_after_retraction.iterrows():
    df_analysis_row = get_df_analise_section.get_df(row.DOI_cited, [row.article_citing], 3) # We will take 3 sentences
    df_3_sentences = df_3_sentences.append(df_analysis_row)

df_3_sentences = df_3_sentences.reset_index(drop = True)
df_3_sentences.tail()

In [ ]:
df_3_sentences

In [ ]:
df_3_sentences[df_3_sentences.cited_DOI == "10.1186/s13033-017-0143-5"]["sentence_citing_maintext"].iloc[0]

In [ ]:
df[df.cited_DOI == "10.1186/s13033-017-0143-5"]["sentence_citing_maintext"].iloc[0]

# USING NLTK


In [ ]:
df_sentences_classification3.sentences_citing.iloc[2]

In [ ]:

df_sentences_classification2.sentences_citing.iloc[2]

In [ ]:
df_sentences_classification.sentences_citing.iloc[2]

In [ ]:
from nltk.corpus import stopwords
for word in stopwords.words('english'):
    print(word)

* but, while, against, not, doesn't ... are stopwords. I think we should keep them.

In [ ]:
import nltk
nltk.download('wordnet')


from nltk.stem import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()

print(lemmatizer.lemmatize("cats"))


In [ ]:
nltk.download('movie_reviews')
 
import nltk   
import random
from nltk.corpus import movie_reviews

documents = [(list(movie_reviews.words(fileid)), category)
             for category in movie_reviews.categories()
             for fileid in movie_reviews.fileids(category)]

random.shuffle(documents)

In [ ]:
movie_reviews.categories()

In [ ]:
documents[2]